In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('E:\\Notes\\Statistics\\Feature Engineering\\Datasets\\titanic_toy.csv')
df.head()

Cabin            Ticket number  Survived
0   NaN         A/5 21171      5         0
1   C85          PC 17599      3         1
2   NaN  STON/O2. 3101282      6         1
3  C123            113803      3         1
4   NaN            373450      A         0

Rows contains different types of Numerical & Categorical Data
- Cabin contains - C85, C123
- Ticket contains - A/5 21171, PC 17599
- number - 5,3,A,6

In [ ]:
# First Transforming Number column - 
df['number_num'] = pd.to_numeric(df['number'], errors='coerce', downcast='integer')
df['number_cat'] = np.where(df['number_num'].isnull(), df['number'], np.nan)

# Filling Missing Values
df['number_num'].fillna(0, inplace=True)
df['number_cat'].fillna('Missing', inplace=True)

df.head()

C:\Users\Aryan\AppData\Local\Temp\ipykernel_28488\4267620103.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['number_num'].fillna(int(0), inplace=True)
C:\Users\Aryan\AppData\Local\Temp\ipykernel_28488\4267620103.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For exampl

Cabin            Ticket number  Survived  number_num number_cat
0   NaN         A/5 21171      5         0         5.0    Missing
1   C85          PC 17599      3         1         3.0    Missing
2   NaN  STON/O2. 3101282      6         1         6.0    Missing
3  C123            113803      3         1         3.0    Missing
4   NaN            373450      A         0         0.0          A

In [16]:
# Transforming Cabin Column
df['cabin_num'] = df['Cabin'].str.extract(r'(\d+)')
df['cabin_cat'] = df['Cabin'].str.extract(r'([A-Za-z]+)')

df.head()

Cabin            Ticket number  Survived  number_num number_cat cabin_num  \
0   NaN         A/5 21171      5         0         5.0    Missing       NaN   
1   C85          PC 17599      3         1         3.0    Missing        85   
2   NaN  STON/O2. 3101282      6         1         6.0    Missing       NaN   
3  C123            113803      3         1         3.0    Missing       123   
4   NaN            373450      A         0         0.0          A       NaN   

  cabin_cat  
0       NaN  
1         C  
2       NaN  
3         C  
4       NaN

In [ ]:
# Transforming Ticket Column - A/5 21171, 113803
df['ticket_num'] = df['Ticket'].apply(lambda s: s.split()[-1]) #Extracts the last nos.
df['ticket_num'] = pd.to_numeric(df['ticket_num'], errors='coerce', downcast='integer')


df['ticket_cat'] = df['Ticket'].apply(lambda s: s.split()[0]) #Extracts the first part
df['ticket_cat'] = np.where(df['ticket_cat'].str.isdigit(), np.nan, df['ticket_cat']) #If the rpw only has digit so cat mai nan se replace ho jayega

df.head(10)

Cabin            Ticket number  Survived  number_num number_cat cabin_num  \
0   NaN         A/5 21171      5         0         5.0    Missing       NaN   
1   C85          PC 17599      3         1         3.0    Missing        85   
2   NaN  STON/O2. 3101282      6         1         6.0    Missing       NaN   
3  C123            113803      3         1         3.0    Missing       123   
4   NaN            373450      A         0         0.0          A       NaN   
5   NaN            330877      2         0         2.0    Missing       NaN   
6   E46             17463      2         0         2.0    Missing        46   
7   NaN            349909      5         0         5.0    Missing       NaN   
8   NaN            347742      1         1         1.0    Missing       NaN   
9   NaN            237736      A         1         0.0          A       NaN   

  cabin_cat  ticket_num ticket_cat  
0       NaN     21171.0        A/5  
1         C     17599.0         PC  
2       NaN   3101282.0   STON/O2.  
3         C    113803.0        NaN  
4       NaN    373450.0        NaN  
5       NaN    330877.0        NaN  
6         E     17463.0        NaN  
7       NaN    349909.0        NaN  
8       NaN    347742.0        NaN  
9       NaN    237736.0        NaN